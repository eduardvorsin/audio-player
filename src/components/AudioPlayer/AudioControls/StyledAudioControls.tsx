import styled from "styled-components";
import { StyledAdditionalControls } from "../AdditionalControls/StyledAdditionalControls";

interface StyledAudioControlsProps {
	className?: string
};

export const StyledAudioControls = styled.div<StyledAudioControlsProps>`
	display: grid;
	justify-content: center;
	gap:15px;
	grid-template-columns: 1fr;
	grid-template-rows: min-content;
	grid-auto-rows: min-content;

	@media(min-width:769px){
		grid-template-columns: 1fr 1fr;
	}

	& > ${StyledAdditionalControls}{
		grid-column: span 2;
	}
`;

