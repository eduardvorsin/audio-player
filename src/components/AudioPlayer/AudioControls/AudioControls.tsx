import React from "react";
import { useScreenWidth } from "../../../hooks/useScreenWidth";
import { StyledAudioControls } from "./StyledAudioControls";
import { MainControls } from "../MainControls/MainControls";
import { AdditionalControls } from "../AdditionalControls/AdditionalControls";
import { VolumeControls } from "../VolumeControls/VolumeControls";

interface AudioControlsProps {
	isPlayed: boolean,
	isMuted: boolean,
	volume?: number,
	showDownloadControl?: boolean,
	showNextAndPreviousControls?: boolean,
	showLoopControl?: boolean,
	showPlaybackRateControl?: boolean,
	isLooped?: boolean,
	playbackRate?: number,
	className?: string,
	downloadLink?: string,
	toggleMuting?: React.MouseEventHandler<HTMLButtonElement>,
	onVolumeChange?: React.ChangeEventHandler<HTMLInputElement>,
	changePlaybackRate?: React.MouseEventHandler<HTMLButtonElement>,
	changeLooping?: React.MouseEventHandler<HTMLButtonElement>,
	togglePlaying?: React.MouseEventHandler<HTMLButtonElement>,
	onClickPrevious?: React.MouseEventHandler<HTMLButtonElement>,
	onClickNext?: React.MouseEventHandler<HTMLButtonElement>,
};

export const AudioControls = React.memo<AudioControlsProps>(({
	isPlayed,
	isMuted,
	volume = 50,
	showDownloadControl = false,
	showPlaybackRateControl = false,
	showNextAndPreviousControls = false,
	showLoopControl = false,
	isLooped = false,
	downloadLink = '',
	playbackRate = 1,
	toggleMuting = () => { },
	onVolumeChange = () => { },
	changeLooping = () => { },
	changePlaybackRate = () => { },
	togglePlaying = () => { },
	onClickPrevious = () => { },
	onClickNext = () => { },
}) => {

	const screenWidth = useScreenWidth();
	const isMobileWidth = screenWidth < 769;
	const showAdditionalControls =
		showDownloadControl || showPlaybackRateControl || showLoopControl;

	return (
		<StyledAudioControls
			data-testid='audio-controls'
		>
			<MainControls
				isPlayed={isPlayed}
				togglePlaying={togglePlaying}
				showNextAndPreviousControls={showNextAndPreviousControls}
				onClickPrevious={onClickPrevious}
				onClickNext={onClickNext}
			/>

			{!isMobileWidth &&
				<VolumeControls
					volume={volume}
					onVolumeChange={onVolumeChange}
					isMuted={isMuted}
					toggleMuting={toggleMuting}
				/>
			}

			{showAdditionalControls &&
				<AdditionalControls
					showDownloadControl={showDownloadControl}
					downloadLink={downloadLink}
					showPlaybackRateControl={showPlaybackRateControl}
					changePlaybackRate={changePlaybackRate}
					playbackRate={playbackRate}
					showLoopControl={showLoopControl}
					changeLooping={changeLooping}
					isLooped={isLooped}
				/>
			}
		</StyledAudioControls>
	);
});