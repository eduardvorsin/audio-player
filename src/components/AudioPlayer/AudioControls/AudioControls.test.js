import { render, screen } from "@testing-library/react";
import { AudioControls } from "./AudioControls";

describe('AudioControls component tests', () => {
  afterEach(() => {
    window.innerWidth = 1024;
  });

  test('renders correctly', () => {
    render(
      <AudioControls
        showDownloadControl
        showPlaybackRateControl
        showLoopControl
      />
    );

    expect(screen.getAllByRole('button')).toHaveLength(5);
  });

  test('renders without passed props', () => {
    render(
      <AudioControls />
    );

    expect(screen.queryAllByRole('button')).toHaveLength(2);
  });

  test('when the screen size is less than 769 px volume controls are not rendered', () => {
    window.innerWidth = 576;
    render(
      <AudioControls />
    );

    expect(screen.queryByRole('button', { name: 'mute' })).not.toBeInTheDocument();
  });

  test('snapshot with default prop values', () => {
    render(
      <AudioControls />
    );

    expect(screen.getByTestId('audio-controls')).toMatchSnapshot();
  });

  test('snapshot with isPlayed & isMuted props', () => {
    render(
      <AudioControls
        isPlayed
        isMuted
      />
    );

    expect(screen.getByTestId('audio-controls')).toMatchSnapshot();
  });

  test('snapshot with a modified volume prop', () => {
    render(
      <AudioControls
        volume={33}
      />
    );

    expect(screen.getByTestId('audio-controls')).toMatchSnapshot();
  });

  test('snapshot with showDownloadControl & downloadLink props', () => {
    render(
      <AudioControls
        showDownloadControl
        downloadLink='test.mp3'
      />
    );

    expect(screen.getByTestId('audio-controls')).toMatchSnapshot();
  });

  test('snapshot with showPlaybackRateControl prop', () => {
    render(
      <AudioControls
        showPlaybackRateControl
      />
    );

    expect(screen.getByTestId('audio-controls')).toMatchSnapshot();
  });

  test('snapshot with showPlaybackRateControl & playbackRate props', () => {
    render(
      <AudioControls
        showPlaybackRateControl
        playbackRate={1.25}
      />
    );

    expect(screen.getByTestId('audio-controls')).toMatchSnapshot();
  });

  test('snapshot with showLoopControl prop', () => {
    render(
      <AudioControls
        showLoopControl
      />
    );

    expect(screen.getByTestId('audio-controls')).toMatchSnapshot();
  });

  test('snapshot with showLoopControl & isLooped props', () => {
    render(
      <AudioControls
        showLoopControl
        isLooped
      />
    );

    expect(screen.getByTestId('audio-controls')).toMatchSnapshot();
  });

  test('snapshot with showNextAndPreviousControls prop', () => {
    render(
      <AudioControls
        showNextAndPreviousControls
      />
    );

    expect(screen.getByTestId('audio-controls')).toMatchSnapshot();
  });

  test('snapshot with all the props affecting the UI', () => {
    render(
      <AudioControls
        showDownloadControl
        showPlaybackRateControl
        showLoopControl
        showNextAndPreviousControls
        className='test'
      />
    );

    expect(screen.getByTestId('audio-controls')).toMatchSnapshot();
  });
});
