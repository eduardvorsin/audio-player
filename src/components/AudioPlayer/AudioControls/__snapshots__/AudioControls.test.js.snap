// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioControls component tests snapshot with a modified volume prop 1`] = `
<div
  class="sc-gswNZR ipfQzw"
  data-testid="audio-controls"
>
  <div
    class="sc-dkrFOg gMtPZj"
    data-testid="main-controls"
  >
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      play
    </button>
  </div>
  <div
    class="sc-jSUZER jBXoGo"
    data-testid="volume-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      mute
    </button>
    <div
      class="sc-gKPRtg mrZYI"
      data-testid="range-slider"
    >
      <input
        style="--progress-percent: 33%;"
        type="range"
        value="33"
      />
    </div>
  </div>
</div>
`;

exports[`AudioControls component tests snapshot with all the props affecting the UI 1`] = `
<div
  class="sc-gswNZR ipfQzw"
  data-testid="audio-controls"
>
  <div
    class="sc-dkrFOg ccDDgC"
    data-testid="main-controls"
  >
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      prev track
    </button>
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      play
    </button>
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      next track
    </button>
  </div>
  <div
    class="sc-jSUZER jBXoGo"
    data-testid="volume-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      mute
    </button>
    <div
      class="sc-gKPRtg mrZYI"
      data-testid="range-slider"
    >
      <input
        style="--progress-percent: 50%;"
        type="range"
        value="50"
      />
    </div>
  </div>
  <div
    class="sc-bcXHqe fDkhxT"
    data-testid="additional-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
      download=""
      href=""
    >
      <svg />
      download track
    </button>
    <button
      aria-label="current playback rate"
      class="sc-hLBbgP bbygWe sc-eDvSVe eCTlAt"
      data-testid="button"
    >
      1x
    </button>
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      repeat song infinite
    </button>
  </div>
</div>
`;

exports[`AudioControls component tests snapshot with default prop values 1`] = `
<div
  class="sc-gswNZR ipfQzw"
  data-testid="audio-controls"
>
  <div
    class="sc-dkrFOg gMtPZj"
    data-testid="main-controls"
  >
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      play
    </button>
  </div>
  <div
    class="sc-jSUZER jBXoGo"
    data-testid="volume-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      mute
    </button>
    <div
      class="sc-gKPRtg mrZYI"
      data-testid="range-slider"
    >
      <input
        style="--progress-percent: 50%;"
        type="range"
        value="50"
      />
    </div>
  </div>
</div>
`;

exports[`AudioControls component tests snapshot with isPlayed & isMuted props 1`] = `
<div
  class="sc-gswNZR ipfQzw"
  data-testid="audio-controls"
>
  <div
    class="sc-dkrFOg gMtPZj"
    data-testid="main-controls"
  >
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      pause
    </button>
  </div>
  <div
    class="sc-jSUZER jBXoGo"
    data-testid="volume-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      unmute
    </button>
    <div
      class="sc-gKPRtg mrZYI"
      data-testid="range-slider"
    >
      <input
        style="--progress-percent: 50%;"
        type="range"
        value="50"
      />
    </div>
  </div>
</div>
`;

exports[`AudioControls component tests snapshot with showDownloadControl & downloadLink props 1`] = `
<div
  class="sc-gswNZR ipfQzw"
  data-testid="audio-controls"
>
  <div
    class="sc-dkrFOg gMtPZj"
    data-testid="main-controls"
  >
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      play
    </button>
  </div>
  <div
    class="sc-jSUZER jBXoGo"
    data-testid="volume-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      mute
    </button>
    <div
      class="sc-gKPRtg mrZYI"
      data-testid="range-slider"
    >
      <input
        style="--progress-percent: 50%;"
        type="range"
        value="50"
      />
    </div>
  </div>
  <div
    class="sc-bcXHqe fDkhxT"
    data-testid="additional-controls"
  >
    <a
      class="sc-hLBbgP djRQit"
      data-testid="button"
      download=""
      href="test.mp3"
    >
      <svg />
      download track
    </a>
  </div>
</div>
`;

exports[`AudioControls component tests snapshot with showLoopControl & isLooped props 1`] = `
<div
  class="sc-gswNZR ipfQzw"
  data-testid="audio-controls"
>
  <div
    class="sc-dkrFOg gMtPZj"
    data-testid="main-controls"
  >
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      play
    </button>
  </div>
  <div
    class="sc-jSUZER jBXoGo"
    data-testid="volume-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      mute
    </button>
    <div
      class="sc-gKPRtg mrZYI"
      data-testid="range-slider"
    >
      <input
        style="--progress-percent: 50%;"
        type="range"
        value="50"
      />
    </div>
  </div>
  <div
    class="sc-bcXHqe fDkhxT"
    data-testid="additional-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      don't repeat the song
    </button>
  </div>
</div>
`;

exports[`AudioControls component tests snapshot with showLoopControl prop 1`] = `
<div
  class="sc-gswNZR ipfQzw"
  data-testid="audio-controls"
>
  <div
    class="sc-dkrFOg gMtPZj"
    data-testid="main-controls"
  >
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      play
    </button>
  </div>
  <div
    class="sc-jSUZER jBXoGo"
    data-testid="volume-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      mute
    </button>
    <div
      class="sc-gKPRtg mrZYI"
      data-testid="range-slider"
    >
      <input
        style="--progress-percent: 50%;"
        type="range"
        value="50"
      />
    </div>
  </div>
  <div
    class="sc-bcXHqe fDkhxT"
    data-testid="additional-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      repeat song infinite
    </button>
  </div>
</div>
`;

exports[`AudioControls component tests snapshot with showNextAndPreviousControls prop 1`] = `
<div
  class="sc-gswNZR ipfQzw"
  data-testid="audio-controls"
>
  <div
    class="sc-dkrFOg ccDDgC"
    data-testid="main-controls"
  >
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      prev track
    </button>
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      play
    </button>
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      next track
    </button>
  </div>
  <div
    class="sc-jSUZER jBXoGo"
    data-testid="volume-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      mute
    </button>
    <div
      class="sc-gKPRtg mrZYI"
      data-testid="range-slider"
    >
      <input
        style="--progress-percent: 50%;"
        type="range"
        value="50"
      />
    </div>
  </div>
</div>
`;

exports[`AudioControls component tests snapshot with showPlaybackRateControl & playbackRate props 1`] = `
<div
  class="sc-gswNZR ipfQzw"
  data-testid="audio-controls"
>
  <div
    class="sc-dkrFOg gMtPZj"
    data-testid="main-controls"
  >
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      play
    </button>
  </div>
  <div
    class="sc-jSUZER jBXoGo"
    data-testid="volume-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      mute
    </button>
    <div
      class="sc-gKPRtg mrZYI"
      data-testid="range-slider"
    >
      <input
        style="--progress-percent: 50%;"
        type="range"
        value="50"
      />
    </div>
  </div>
  <div
    class="sc-bcXHqe fDkhxT"
    data-testid="additional-controls"
  >
    <button
      aria-label="current playback rate"
      class="sc-hLBbgP bbygWe sc-eDvSVe eCTlAt"
      data-testid="button"
    >
      1.25x
    </button>
  </div>
</div>
`;

exports[`AudioControls component tests snapshot with showPlaybackRateControl prop 1`] = `
<div
  class="sc-gswNZR ipfQzw"
  data-testid="audio-controls"
>
  <div
    class="sc-dkrFOg gMtPZj"
    data-testid="main-controls"
  >
    <button
      class="sc-hLBbgP eWuHc"
      data-testid="button"
    >
      <svg />
      play
    </button>
  </div>
  <div
    class="sc-jSUZER jBXoGo"
    data-testid="volume-controls"
  >
    <button
      class="sc-hLBbgP djRQit"
      data-testid="button"
    >
      <svg />
      mute
    </button>
    <div
      class="sc-gKPRtg mrZYI"
      data-testid="range-slider"
    >
      <input
        style="--progress-percent: 50%;"
        type="range"
        value="50"
      />
    </div>
  </div>
  <div
    class="sc-bcXHqe fDkhxT"
    data-testid="additional-controls"
  >
    <button
      aria-label="current playback rate"
      class="sc-hLBbgP bbygWe sc-eDvSVe eCTlAt"
      data-testid="button"
    >
      1x
    </button>
  </div>
</div>
`;
